src/startup.c

```diff
@@ -134,7 +134,23 @@ void dump_boot_args(struct boot_args *ba)
             break;
     }
     if (!mem_size_actual) {
-        mem_size_actual = ALIGN_UP(ba->phys_base + ba->mem_size - 0x800000000, BIT(30));
+        if (chip_id == T8012) {
+            int anode = adt_path_offset(adt, "/arm-io/mcc");
+
+            /*
+             * Lower 512 MB intented for OS use, upper 512 or 1536 MB is some sort
```

Comment:
Aren't we the OS? Or is there something else running there? And I'm not sure how this comments related to the code below which just computes the memory size.

The SSD cache is also present on >M1 fwiw, it's transparently handled by ANS firmware there and we can't even access it from the AP.

Reply:
The comment help explain why the amount of reserved memory is so big. The upper memory region has something to do with exporting the ANS2 storage to the x86 CPU (which, on a T2 machine, is more like a big fat coprocessor of the T2).

Reply:
I think that upper memory is probably just Apple's variant of host memory buffers, i.e. using the main RAM as cache for the nvme drive. I don't even think that will need any driver support as long as you make sure not to clobber that memory region. When I briefly looked at how NVMe worked on T2 macs it seemed like it's just exporting the same MMIO interface we directly use on >M1 to the x86 CPU.

What I'm confused about here is that this comment talks about two reserved memory regions at different locations (lower and upper) but then the code just computes mem_size_actual.